Library/Homebrew/migrator.rb

```diff
@@ -0,0 +1,327 @@
+
+    @old_tabs = oldpath.subdirs.each.map { |d| Tab.for_keg(Keg.new(d)) }
+    @old_tap = old_tabs.first.tap
+    raise MigratorDifferentTapsError.new(formula, old_tap) unless from_same_taps?
+
+    @newpath = HOMEBREW_CELLAR/formula.name
+    @oldname = formula.oldname
+    @newname = formula.name
+
+    if @oldkeg = get_linked_oldkeg
+      @old_linked_keg_record = oldkeg.linked_keg_record if oldkeg.linked?
+      @old_opt_record = oldkeg.opt_record if oldkeg.optlinked?
+    end
+
+    @old_pin_record = HOMEBREW_LIBRARY/"PinnedKegs"/oldname
+    @new_pin_record = HOMEBREW_LIBRARY/"PinnedKegs"/newname
+    @pinned = old_pin_record.symlink?
+    @old_pin_link_record = old_pin_record.readlink if @pinned
+  end
+
+  # Check if installed keg is from core formulae and newname is from core.
+  def both_from_core?
+    new_core = formula.core_formula?
+    old_core = (old_tap.nil? || old_tap == "Homebrew/homebrew" \
+      || old_tap == "mxcl/master")
+    # Formula can be installed from path and we have to check wheter this path
+    # is the same as for core formula.
+    if (path = old_tabs.first.source["path"]) \
+        && path != Formulary.core_path(formula.oldname).to_s
+      old_core = false
+    end
+    new_core && old_core
+  end
+
+  # Fix INSTALL_RECEIPTS for tap-migrated formula.
+  def fix_tabs
+    old_tabs.each do |tab|
+      tab.source["tap"] = formula.tap
+      tab.write
+    end
+  end
+
+  def from_same_taps?
+    if both_from_core?
+      true
+    elsif formula.tap == old_tap
+      true
+    # Homebrew didn't use to update tabs while performing tap-migrations,
+    # so there can be INSTALL_RECEIPT's containing wrong information about
+    # tap (tap is Homebrew/homebrew if installed formula migrates to a tap), so
+    # we check if there is an entry about oldname migrated to tap and if
+    # newname's tap is the same as tap to which oldname migrated, then we
+    # can perform migrations and the taps for oldname and newname are the same.
+    else
+      require "tap_migrations"
+      if TAP_MIGRATIONS && (rec = TAP_MIGRATIONS[formula.oldname]) \
+          && rec.sub("/", "/homebrew-") == formula.tap
+        fix_tabs
+        true
+      else
+        false
+      end
+    end
+  end
+
+  def get_linked_oldkeg
+    kegs = oldpath.subdirs.map { |d| Keg.new(d) }
+    kegs.detect(&:linked?) || kegs.detect(&:optlinked?)
+  end
+
+  def pinned?
+    @pinned
+  end
+
+  def oldkeg_linked?
+    !!oldkeg
+  end
+
+  def migrate
+    if newpath.exist?
+      opoo "#{newpath} already exists; remove it manually and run brew migrate #{oldname}."
+    else
+      begin
+        oh1 "Migrating #{Tty.green}#{oldname}#{Tty.white} to #{Tty.green}#{newname}#{Tty.reset}"
+        copy_to_new_directory
+        repin
+        remove_and_unlink_oldname
+        link_newname
+        link_oldname_opt
+        link_oldname_cellar
+        update_tabs
+      # TODO
+      # - Make some output about exception.
+      rescue Exception
```

Comment:
It may just be worth doing `rescue` without naming any exceptions to handle all unconditionally. `formula_versions` names specific exceptions because those are the only ones that occur in those circumstances whereas here we really do want to catch every possible exception.